### PR TITLE
Asynchrounous refactoring

### PR DESCRIPTION
**Pull request by Trivernis**

### CODE REVIEW

**Comment by Trivernis** [comment]
Maybe I should just merge this into my forked main and do some heavy refactoring before merging

**Comment by not-my-segfault** [comment]
> Maybe I should just merge this into my forked main and do some heavy refactoring before merging into upstream

Whichever works best for your workflow, we've currently got changes to the search feature on the development branch, but that's the only difference to main, so it might be a good idea to open up the PR against development.

I'm not gonna be doing any more work on Amethyst for now as I'm on holiday, so there shouldn't be much more extra work from doing so

**Comment by Trivernis** [comment]
Nice, more conflicts

**Comment by not-my-segfault** [comment]
Don't mind me just making sure things are tagged correctly

**Comment by Trivernis** [comment]
I'm tired. Someone look at this and tell me it works

**Comment by not-my-segfault** [comment]
> I'm tired. Someone look at this and tell me it works

I'm on my way!

**Comment by not-my-segfault** [comment]
One issue I've noticed so far is that the indentation/word wrapping seems to be off in some places (I'm installing `aws-cli-git` as one of the examples provided by the [Arch Wiki page](https://wiki.archlinux.org/title/AUR_helpers#Legend)

image

**Comment by Trivernis** [comment]
> One issue I've noticed so far is that the indentation/word wrapping seems to be off in some places (I'm installing `aws-cli-git` as one of the examples provided by the [Arch Wiki page](https://wiki.archlinux.org/title/AUR_helpers#Legend)
> 
> image

lmao yea I just broke it

**Comment by not-my-segfault** [comment]
~~`aws-cli-git` actually fails to install, this could be a result of it's non-standard PKGBUILD, worth looking into potentially?~~

It turns out not even `makepkg` seems to be able to handle this package. Lmao

image

**Comment by not-my-segfault** [comment]
image
I'm not entirely sure what happened here?

**Comment by not-my-segfault** [comment]
> image I'm not entirely sure what happened here?

image

This is the output with `-vv`, super nondescriptive

**Comment by Trivernis** [comment]
> > image I'm not entirely sure what happened here?
> 
> image
> 
> This is the output with `-vv`, super nondescriptive, is it potentially a PKGBUILD issue?

Exit code 9 would be RpcError tho. So something with the aur rpc

**Comment by not-my-segfault** [comment]
I've noticed from all of this that the wiki page for AUR helpers is riddled with terrible PKGBUILDs, please for the love of god take a look at whatever the hell this is https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=aws-cli-git

**Comment by Trivernis** [comment]
> I've noticed from all of this that the wiki page for AUR helpers is riddled with terrible PKGBUILDs, please for the love of god take a look at whatever the hell this is https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=aws-cli-git

God is dead

**Comment by axtloss** [comment]
> I've noticed from all of this that the wiki page for AUR helpers is riddled with terrible PKGBUILDs, please for the love of god take a look at whatever the hell this is https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=aws-cli-git

I'm going to cry

**Comment by SomethingGeneric** [comment]
and you all complain about *my* pkgbuilds lmao

**Comment by SomethingGeneric** [comment]
this is horrid

**Comment by Trivernis** [comment]
The ros-melodic-desktop error is caused by a weird rpc response that cannot be parsed by serde. I have to fix that in the rpc wrapper pkg.
Serde didn't expect the Maintainer field to be null.

**Comment by not-my-segfault** [comment]
Maybe supporting the AUR was a mistake. I'm going to retreat into a cave forever now, goodbye all! /j

**Comment by not-my-segfault** [comment]
On a positive note, the new PKGBUILD parsing code seems to handle split pacakges beautifully! CLion was a super simple install process.

**Comment by not-my-segfault** [comment]
Potentially terrible idea, we can actually handle shell completions for all AUR packages, as a list of all packages in the AUR is available as a GZIP-compressed plaintext file at https://aur.archlinux.org/packages.gz

I might experiment with how awful this would be to implement at some point

**Comment by not-my-segfault** [comment]
One issue, when installing split packages like Clion (Which pulls in all of it's optdepends), every optdepend is installed as explicit. This can lead to cluttering the system with unnecessary packages

**Comment by not-my-segfault** [comment]
Also a formatting thing, the `No pacnew files found` text should be bold

image

**Comment by Trivernis** [comment]
> Potentially terrible idea, we can actually handle shell completions for all AUR packages, as a list of all packages in the AUR is available as a GZIP-compressed plaintext file at https://aur.archlinux.org/packages.gz
> 
> I might experiment with how awful this would be to implement at some point

lmao. We can just download that and put it in ~/.cache ig

**Comment by Trivernis** [comment]
> One issue, when installing split packages like Clion (Which pulls in all of it's optdepends), every optdepend is installed as explicit. This can lead to cluttering the system with unnecessary packages

optdepends shouldn't be installed automatically

**Comment by Trivernis** [comment]
Hmm, looks like pacman is doing it

**Comment by Trivernis** [comment]
@axtloss @not-my-segfault I *slightly* refactored the aur install and split it into several files.
It should be readable now. I'd like to merge this (before it gets too big) if there's no other issues that weren't mentioned yet.

**Comment by not-my-segfault** [comment]
Makes sense, I'll be free in around an hour to test it a bit further. Thanks for all the good work!

**Comment by axtloss** [comment]
Yeah seems fine, although it should probably squash and merge to not clutter the git history

**Comment by Trivernis** [comment]
> Yeah seems fine, although it should probably squash and merge to not clutter the git history

Squashing is kinda risky because you can quickly run into merge conflicts by doing that as it rewrites the history.
I had nothing but pain with squashing in the past.

**Comment by axtloss** [comment]
oh then just a regular merge ig

**Comment by not-my-segfault** [comment]
> Yeah seems fine, although it should probably squash and merge to not clutter the git history

We'll rebase and merge, that's part of our new Git workflow

File: Cargo.toml
Comment (not-my-segfault) [inline Cargo.toml:12]: Some of the Cargo.toml file seems to have bit the dust, we no longer have the pkg-warner feature and i'd like to keep the `repository`, `keywords`, and `categories` keywords, so we can push to crates.io at some point

**Review by not-my-segfault — CHANGES_REQUESTED** [review]

**Review by not-my-segfault — APPROVED** [review]

**Comment by not-my-segfault** [comment]
There seem to be some merge issues that GitHub isn't telling us about for rebasing and committing. Sigh, I'll see what's up in Git lol

**Comment by Trivernis** [comment]
> There seem to be some merge issues that GitHub isn't telling us about for rebasing and committing. Sigh, I'll see what's up in Git lol

Yea rebasing is witchcraft

**Comment by not-my-segfault** [comment]
I'll manually rebase and push to `development`, thanks for all the great work again :)

**Comment by Trivernis** [comment]
lol

**Pull request merged**
